types/src/diem_id_identifier.rs

```diff
@@ -0,0 +1,274 @@
+            ));
+        }
+
+        let mut chars = s.chars();
+        match chars.next() {
+            Some('a'..='z') | Some('A'..='Z') | Some('0'..='9') => {}
+            Some(_) => {
+                return Err(DiemIdParseError::new(
+                    "Invalid user identifier input".to_string(),
+                ))
+            }
+            None => {
+                return Err(DiemIdParseError::new(
+                    "Invalid user identifier input".to_string(),
+                ))
+            }
+        }
+        for c in chars {
+            match c {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '.' => {}
+                _ => return Err(DiemIdParseError::new(format!("Invalid character {}", c))),
+            }
+        }
+        Ok(DiemIdUserIdentifier(s.into()))
+    }
+
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+impl<'de> de::Deserialize<'de> for DiemIdUserIdentifier {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let s = <String>::deserialize(deserializer)?;
+        DiemIdUserIdentifier::new(&s).map_err(D::Error::custom)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct DiemIdVaspDomainIdentifier(Box<str>);
+
+impl DiemIdVaspDomainIdentifier {
+    pub fn new(s: &str) -> Result<Self, DiemIdParseError> {
+        if s.len() > 63 || s.is_empty() {
+            return Err(DiemIdParseError::new(
+                "Invalid length for vasp domain identifier".to_string(),
+            ));
+        }
+
+        let mut chars = s.chars();
+        match chars.next() {
+            Some('a'..='z') | Some('A'..='Z') | Some('0'..='9') => {}
+            Some(_) => {
+                return Err(DiemIdParseError::new(
+                    "Invalid vasp domain input".to_string(),
+                ))
+            }
+            None => {
+                return Err(DiemIdParseError::new(
+                    "Invalid vasp domain input".to_string(),
+                ))
+            }
+        }
+        for c in chars {
+            match c {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '.' => {}
+                _ => return Err(DiemIdParseError::new(format!("Invalid character {}", c))),
+            }
+        }
+        Ok(DiemIdVaspDomainIdentifier(s.into()))
+    }
+
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+impl<'de> de::Deserialize<'de> for DiemIdVaspDomainIdentifier {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let s = <String>::deserialize(deserializer)?;
+        DiemIdVaspDomainIdentifier::new(&s).map_err(D::Error::custom)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DiemIdParseError {
+    message: String,
+}
+
+impl DiemIdParseError {
+    pub fn new(message: String) -> DiemIdParseError {
```

Comment:
Let's not mark this as public (as a way to hide the implementation details of the Error type)